src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/PkcsPalOpenSsl.Encrypt.cs

```diff
@@ -70,14 +71,94 @@ internal sealed partial class PkcsPalOpenSsl : PkcsPal
+        {
+            DerSequenceReader originalMessage = new DerSequenceReader(encryptedMessage);
+
+            // EnvelopedCms has the following underlying structure:
+            //
+            // ContentInfo ::= SEQUENCE {
+            //      contentType ContentType,
+            //      content[0] EXPLICIT ANY DEFINED BY contentType }
+            //
+            // ContentType ::= OBJECT IDENTIFIER
+            //
+            // According to RFC2630, the enveloped type we have support for has the following structure
+            // EnvelopedData ::= SEQUENCE {
+            //     version CMSVersion,
+            //     originatorInfo[0] IMPLICIT OriginatorInfo OPTIONAL,
+            //     recipientInfos RecipientInfos,
+            //     encryptedContentInfo EncryptedContentInfo,
+            //     unprotectedAttrs[1] IMPLICIT UnprotectedAttributes OPTIONAL }
+
+            byte[][] contentType = originalMessage.ReadAndSplitNextEncodedValue();
+            DerSequenceReader envelopedDataReader = originalMessage.ReadExplicitContextSequence();
+
+            byte[][] version = envelopedDataReader.ReadAndSplitNextEncodedValue();
+
+            const byte ConstructedContextSpecificTag = DerSequenceReader.ContextSpecificTagFlag | DerSequenceReader.ConstructedFlag;
+
+            byte[][] originatorInfo = (envelopedDataReader.PeekTag() == ConstructedContextSpecificTag) ?
+                originatorInfo = envelopedDataReader.ReadAndSplitNextEncodedValue() :
+                new byte[][] { Array.Empty<byte>(), Array.Empty<byte>(), Array.Empty<byte>() };
+
+            byte[][] recipientInfos = envelopedDataReader.ReadAndSplitNextEncodedValue();
+            byte[][] encryptedContentInfo = envelopedDataReader.ReadAndSplitNextEncodedValue();
+
+            byte[][] envelopedData = DerEncoder.ConstructSegmentedExplicitSequence(
+                0 /* Context number */,
+                version,
+                originatorInfo,
+                recipientInfos,
+                encryptedContentInfo,
+                EncodeAttributes(unprotectedAttributes));
+
+            return DerEncoder.ConstructSequence(
+                contentType,
+                envelopedData);
+        }
+
+        private byte[][] EncodeAttributes(CryptographicAttributeObjectCollection unprotectedAttributes)
+        {
+            // In EnvelopedData UnprotectedAttributes are included as:
+            //
+            // unprotectedAttrs[1] IMPLICIT UnprotectedAttributes OPTIONAL
+            //
+            // UnprotectedAttributes ::= SET SIZE (1..MAX) OF Attribute
+            // Attribute::= SEQUENCE {
+            //      attrType OBJECT IDENTIFIER,
+            //      attrValues SET OF AttributeValue }
+
+            byte[][][] setOfAttrs = new byte[unprotectedAttributes.Count][][];
+            
+            for (int i = 0; i < unprotectedAttributes.Count; i++)
```

Comment:
This needs a lot of explanation of what's going on.  And has a significant amount of array exploding and reassembling.  Is all of it necessary?

Do you guarantee that the data is well-ordered?

> The encodings of the component values of a set-of value shall appear in ascending order, the encodings being compared as octet strings with the shorter components being padded at their trailing end with 0-octets.

(X.690 11.6: Set-of components)